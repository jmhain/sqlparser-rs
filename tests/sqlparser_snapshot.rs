// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#![warn(clippy::all)]

use sqlparser::ast::Statement;
use sqlparser::dialect::{Dialect, GenericDialect};
use sqlparser::parser::Parser;
use std::path::Path;

fn test_dialect(path: &Path, dialect: &dyn Dialect) -> datatest_stable::Result<()> {
    let sql = std::fs::read_to_string(path)?;

    let stmts = Parser::parse_sql(dialect, &sql)?;
    let name = path
        .strip_prefix(Path::new("tests/queries"))?
        .to_str()
        .unwrap();
    insta::assert_yaml_snapshot!(name, stmts);

    // Rendering the AST to a string and re-parsing it should yield the same AST.
    for stmt in stmts {
        let [round_trip_stmt]: [Statement; 1] = Parser::parse_sql(dialect, &stmt.to_string())?
            .try_into()
            .map_err(|_| "expected a single statement")?;
        pretty_assertions::assert_eq!(stmt, round_trip_stmt);
    }

    Ok(())
}

fn test_generic(path: &Path) -> datatest_stable::Result<()> {
    test_dialect(path, &GenericDialect)
}

datatest_stable::harness!(test_generic, "tests/queries/generic", r"^.*.sql",);
