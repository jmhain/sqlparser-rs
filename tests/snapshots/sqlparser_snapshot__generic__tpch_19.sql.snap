---
source: tests/sqlparser_snapshot.rs
expression: stmts
---
- Query:
    with: ~
    body:
      Select:
        distinct: ~
        top: ~
        projection:
          - ExprWithAlias:
              expr:
                Function:
                  name:
                    - value: sum
                      quote_style: ~
                  args:
                    - Unnamed:
                        Expr:
                          BinaryOp:
                            left:
                              Identifier:
                                value: l_extendedprice
                                quote_style: ~
                            op: Multiply
                            right:
                              Nested:
                                BinaryOp:
                                  left:
                                    Value:
                                      Number:
                                        - "1"
                                        - false
                                  op: Minus
                                  right:
                                    Identifier:
                                      value: l_discount
                                      quote_style: ~
                  filter: ~
                  null_treatment: ~
                  over: ~
                  distinct: false
                  special: false
                  order_by: []
              alias:
                value: revenue
                quote_style: ~
        into: ~
        from:
          - relation:
              Table:
                name:
                  - value: lineitem
                    quote_style: ~
                alias: ~
                args: ~
                with_hints: []
                version: ~
                partitions: []
            joins: []
          - relation:
              Table:
                name:
                  - value: part
                    quote_style: ~
                alias: ~
                args: ~
                with_hints: []
                version: ~
                partitions: []
            joins: []
        lateral_views: []
        selection:
          BinaryOp:
            left:
              BinaryOp:
                left:
                  Nested:
                    BinaryOp:
                      left:
                        BinaryOp:
                          left:
                            BinaryOp:
                              left:
                                BinaryOp:
                                  left:
                                    BinaryOp:
                                      left:
                                        BinaryOp:
                                          left:
                                            BinaryOp:
                                              left:
                                                BinaryOp:
                                                  left:
                                                    Identifier:
                                                      value: p_partkey
                                                      quote_style: ~
                                                  op: Eq
                                                  right:
                                                    Identifier:
                                                      value: l_partkey
                                                      quote_style: ~
                                              op: And
                                              right:
                                                BinaryOp:
                                                  left:
                                                    Identifier:
                                                      value: p_brand
                                                      quote_style: ~
                                                  op: Eq
                                                  right:
                                                    Value:
                                                      SingleQuotedString: "Brand#12"
                                          op: And
                                          right:
                                            InList:
                                              expr:
                                                Identifier:
                                                  value: p_container
                                                  quote_style: ~
                                              list:
                                                - Value:
                                                    SingleQuotedString: SM CASE
                                                - Value:
                                                    SingleQuotedString: SM BOX
                                                - Value:
                                                    SingleQuotedString: SM PACK
                                                - Value:
                                                    SingleQuotedString: SM PKG
                                              negated: false
                                      op: And
                                      right:
                                        BinaryOp:
                                          left:
                                            Identifier:
                                              value: l_quantity
                                              quote_style: ~
                                          op: GtEq
                                          right:
                                            Value:
                                              Number:
                                                - "1"
                                                - false
                                  op: And
                                  right:
                                    BinaryOp:
                                      left:
                                        Identifier:
                                          value: l_quantity
                                          quote_style: ~
                                      op: LtEq
                                      right:
                                        BinaryOp:
                                          left:
                                            Value:
                                              Number:
                                                - "1"
                                                - false
                                          op: Plus
                                          right:
                                            Value:
                                              Number:
                                                - "10"
                                                - false
                              op: And
                              right:
                                Between:
                                  expr:
                                    Identifier:
                                      value: p_size
                                      quote_style: ~
                                  negated: false
                                  low:
                                    Value:
                                      Number:
                                        - "1"
                                        - false
                                  high:
                                    Value:
                                      Number:
                                        - "5"
                                        - false
                          op: And
                          right:
                            InList:
                              expr:
                                Identifier:
                                  value: l_shipmode
                                  quote_style: ~
                              list:
                                - Value:
                                    SingleQuotedString: AIR
                                - Value:
                                    SingleQuotedString: AIR REG
                              negated: false
                      op: And
                      right:
                        BinaryOp:
                          left:
                            Identifier:
                              value: l_shipinstruct
                              quote_style: ~
                          op: Eq
                          right:
                            Value:
                              SingleQuotedString: DELIVER IN PERSON
                op: Or
                right:
                  Nested:
                    BinaryOp:
                      left:
                        BinaryOp:
                          left:
                            BinaryOp:
                              left:
                                BinaryOp:
                                  left:
                                    BinaryOp:
                                      left:
                                        BinaryOp:
                                          left:
                                            BinaryOp:
                                              left:
                                                BinaryOp:
                                                  left:
                                                    Identifier:
                                                      value: p_partkey
                                                      quote_style: ~
                                                  op: Eq
                                                  right:
                                                    Identifier:
                                                      value: l_partkey
                                                      quote_style: ~
                                              op: And
                                              right:
                                                BinaryOp:
                                                  left:
                                                    Identifier:
                                                      value: p_brand
                                                      quote_style: ~
                                                  op: Eq
                                                  right:
                                                    Value:
                                                      SingleQuotedString: "Brand#23"
                                          op: And
                                          right:
                                            InList:
                                              expr:
                                                Identifier:
                                                  value: p_container
                                                  quote_style: ~
                                              list:
                                                - Value:
                                                    SingleQuotedString: MED BAG
                                                - Value:
                                                    SingleQuotedString: MED BOX
                                                - Value:
                                                    SingleQuotedString: MED PKG
                                                - Value:
                                                    SingleQuotedString: MED PACK
                                              negated: false
                                      op: And
                                      right:
                                        BinaryOp:
                                          left:
                                            Identifier:
                                              value: l_quantity
                                              quote_style: ~
                                          op: GtEq
                                          right:
                                            Value:
                                              Number:
                                                - "10"
                                                - false
                                  op: And
                                  right:
                                    BinaryOp:
                                      left:
                                        Identifier:
                                          value: l_quantity
                                          quote_style: ~
                                      op: LtEq
                                      right:
                                        BinaryOp:
                                          left:
                                            Value:
                                              Number:
                                                - "10"
                                                - false
                                          op: Plus
                                          right:
                                            Value:
                                              Number:
                                                - "10"
                                                - false
                              op: And
                              right:
                                Between:
                                  expr:
                                    Identifier:
                                      value: p_size
                                      quote_style: ~
                                  negated: false
                                  low:
                                    Value:
                                      Number:
                                        - "1"
                                        - false
                                  high:
                                    Value:
                                      Number:
                                        - "10"
                                        - false
                          op: And
                          right:
                            InList:
                              expr:
                                Identifier:
                                  value: l_shipmode
                                  quote_style: ~
                              list:
                                - Value:
                                    SingleQuotedString: AIR
                                - Value:
                                    SingleQuotedString: AIR REG
                              negated: false
                      op: And
                      right:
                        BinaryOp:
                          left:
                            Identifier:
                              value: l_shipinstruct
                              quote_style: ~
                          op: Eq
                          right:
                            Value:
                              SingleQuotedString: DELIVER IN PERSON
            op: Or
            right:
              Nested:
                BinaryOp:
                  left:
                    BinaryOp:
                      left:
                        BinaryOp:
                          left:
                            BinaryOp:
                              left:
                                BinaryOp:
                                  left:
                                    BinaryOp:
                                      left:
                                        BinaryOp:
                                          left:
                                            BinaryOp:
                                              left:
                                                Identifier:
                                                  value: p_partkey
                                                  quote_style: ~
                                              op: Eq
                                              right:
                                                Identifier:
                                                  value: l_partkey
                                                  quote_style: ~
                                          op: And
                                          right:
                                            BinaryOp:
                                              left:
                                                Identifier:
                                                  value: p_brand
                                                  quote_style: ~
                                              op: Eq
                                              right:
                                                Value:
                                                  SingleQuotedString: "Brand#34"
                                      op: And
                                      right:
                                        InList:
                                          expr:
                                            Identifier:
                                              value: p_container
                                              quote_style: ~
                                          list:
                                            - Value:
                                                SingleQuotedString: LG CASE
                                            - Value:
                                                SingleQuotedString: LG BOX
                                            - Value:
                                                SingleQuotedString: LG PACK
                                            - Value:
                                                SingleQuotedString: LG PKG
                                          negated: false
                                  op: And
                                  right:
                                    BinaryOp:
                                      left:
                                        Identifier:
                                          value: l_quantity
                                          quote_style: ~
                                      op: GtEq
                                      right:
                                        Value:
                                          Number:
                                            - "20"
                                            - false
                              op: And
                              right:
                                BinaryOp:
                                  left:
                                    Identifier:
                                      value: l_quantity
                                      quote_style: ~
                                  op: LtEq
                                  right:
                                    BinaryOp:
                                      left:
                                        Value:
                                          Number:
                                            - "20"
                                            - false
                                      op: Plus
                                      right:
                                        Value:
                                          Number:
                                            - "10"
                                            - false
                          op: And
                          right:
                            Between:
                              expr:
                                Identifier:
                                  value: p_size
                                  quote_style: ~
                              negated: false
                              low:
                                Value:
                                  Number:
                                    - "1"
                                    - false
                              high:
                                Value:
                                  Number:
                                    - "15"
                                    - false
                      op: And
                      right:
                        InList:
                          expr:
                            Identifier:
                              value: l_shipmode
                              quote_style: ~
                          list:
                            - Value:
                                SingleQuotedString: AIR
                            - Value:
                                SingleQuotedString: AIR REG
                          negated: false
                  op: And
                  right:
                    BinaryOp:
                      left:
                        Identifier:
                          value: l_shipinstruct
                          quote_style: ~
                      op: Eq
                      right:
                        Value:
                          SingleQuotedString: DELIVER IN PERSON
        group_by:
          Expressions: []
        cluster_by: []
        distribute_by: []
        sort_by: []
        having: ~
        named_window: []
        qualify: ~
    order_by: []
    limit: ~
    limit_by: []
    offset: ~
    fetch: ~
    locks: []

